veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,330 @@
+
+## API Examples
+
+### Enable/disable QEMU backup
+
+A new `QCOW2Overlay` field will be added to the VirtualMachine and VirtualMachineInstance CRs. In the VirtualMachine CR, it will be added under: `Spec.Template.Spec.Domain.Devices.QCOW2Overlay`. Similarly, in VirtualMachineInstance CR, it will be added under: `Spec.Domain.Devices.QCOW2Overlay`.
+The field will be of type \*bool and will determine whether the QCOW2 overlay should be applied. As with any other field in the VMI template, this change will only take effect when the VM is restarted. Upon restart, the VMI spec will be created with the updated field value.
+
+When the field is present and set to `true`, a VM state PVC will be created if doesn't already exist. A QCOW2 image will then be created using the existing raw disk image as its data-file. Before applying the domain XML and starting it in virt-launcher manager, the XML will be modified to use the newly created QCOW2 image as the disk.
+
+The XML will modified from:
+```xml
+    <disk type='file' device='disk' model='virtio-non-transitional'>
+      <driver name='qemu' type='raw' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/var/run/kubevirt-private/vmi-disks/datavolumedisk/disk.img' index='2'/>
+      ...
+    </disk>
+```
+
+To:
+```xml
+    <disk type='file' device='disk' model='virtio-non-transitional'>
+      <driver name='qemu' type='qcow2' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/run/kubevirt-private/libvirt/qemu/swtpm/qcow2overlay.qcow2' index='2'>
+        <dataStore type='file'>
+          <format type='raw'/>
+          <source file='/run/kubevirt-private/vmi-disks/datavolumedisk/disk.img' index='3'/>
+        </dataStore>
+      </source>
+      ...
+    </disk>
+```
+
+In the case the field is not present or set to `false`, the VM state PVC will deleted if it is not required by other features. The QCOW2 image will be deleted if it exists. And the domain XML will be modified back to use the raw image as the disk.
+
+
+### VirtualMachineBackup CRD
+
+`VirtualMachineBackup` is a namespace-scoped Custom Resource that initiates the backup process.
+
+The CR name should be a unique identifier for the backup within the namespace, with only one backup allowed per VM at a time.
+The CR spec includes the following fields:
+- `source`<br>
+Specifying the VM source for the backup.
+- `mode`<br>
+Optional. Should be either `push` or `pull`. Initially, as mentioned, only `push` will be allowed. If not specified, `push` will be the default behavior.
+- `incremental`<br>
+Optional. If not specified, a full backup will be taken. When set, the field value should be the checkpoint name from which the incremental backup should be taken.
+
```

Comment:
Recommending to add a pvcName field here.  For push, it will write the backup data to this PVC (instead of creating a PVC.  For pull mode, this PVC would be used for the scratch space.

How will kubevirt decide to create the PVC if it must?  Which storage class would be used for example?

Reply:
maybe the same as defined for the vm state?

Reply:
Or do you think we should use the storage spec in cdi or something like that?

Reply:
I think it's complicated to get right and it's much easier to leave this task to the backup client.  In the future we could add automatic PVC provisioning.  Choosing a correct size for the PVC will also be harder since it can depend on many factors that only the backup client is privy to (ie. time since last backup, policy, etc).

Reply:
@ShellyKa13 can you please define who is creating this cr? Backup vendor?Guest owner?